diff -up ghostscript-9.07/psi/zfapi.c.zfapi-crash ghostscript-9.07/psi/zfapi.c
--- ghostscript-9.07/psi/zfapi.c.zfapi-crash	2013-07-08 12:27:53.136718093 +0100
+++ ghostscript-9.07/psi/zfapi.c	2013-07-08 12:28:17.241826842 +0100
@@ -2141,6 +2141,8 @@ FAPI_char(i_ctx_t *i_ctx_p, bool bBuildG
         /* initialise the FAPI font, this includes language specific stuff */
         I->ff = ps_ff_stub;
 
+        I->client_ctx_p = i_ctx_p;
+
         if (bBuildGlyph && !bCID) {
             if (r_type(op) != t_name) {
                 name_enter_string(imemory, ".notdef", op);
