diff -up ghostscript-9.07/cups/gstoraster.convs.gstoraster-costs ghostscript-9.07/cups/gstoraster.convs
--- ghostscript-9.07/cups/gstoraster.convs.gstoraster-costs	2013-02-14 07:58:16.000000000 +0000
+++ ghostscript-9.07/cups/gstoraster.convs	2013-08-21 12:52:25.255346352 +0100
@@ -26,5 +26,5 @@
 #
 # CUPS file conversion rules for gstoraster filter
 
-application/vnd.cups-pdf	application/vnd.cups-raster	66	gstoraster
-application/vnd.cups-postscript	application/vnd.cups-raster	100	gstoraster
+application/vnd.cups-pdf	application/vnd.cups-raster	99	gstoraster
+application/vnd.cups-postscript	application/vnd.cups-raster	175	gstoraster
